monkestation/code/modules/replays/subsystem/replay.dm

```diff
@@ -406,32 +408,72 @@ SUBSYSTEM_DEF(demo)
 			overlays_identical ? "" : overlays_string,
 			underlays_identical ? "" :underlays_string
 			)
-		while(diffed_appearance_list[diffed_appearance_list.len] == "" && diffed_appearance_list.len > 0)
+		while(diffed_appearance_list[length(diffed_appearance_list)] == "" && length(diffed_appearance_list) > 0)
 			diffed_appearance_list.len--
 
 		var/diffed_string = "~{[jointext(diffed_appearance_list, ";")]}"
 		if(length(diffed_string) < length(undiffed_string))
 			return diffed_string
 	return undiffed_string
 
-/datum/controller/subsystem/demo/stat_entry(msg)
-	msg += "Remaining: {"
-	msg += "Trf:[marked_turfs.len]|"
-	msg += "New:[marked_new.len]|"
-	msg += "Upd:[marked_dirty.len]|"
-	msg += "Del:[del_list.len]"
-	msg += "}"
-	return ..()
+/datum/controller/subsystem/demo/proc/write_time()
+	var/new_time = world.time
+	if(last_written_time != new_time)
+		if(initialized)
+			WRITE_LOG_NO_FORMAT(GLOB.demo_log, "time [new_time]\n")
+		else
+			pre_init_lines += "time [new_time]"
+	last_written_time = new_time
+
+/datum/controller/subsystem/demo/proc/write_event_line(line)
+	write_time()
+	if(initialized)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "[line]\n")
+	else
+		pre_init_lines += line
+
+/datum/controller/subsystem/demo/proc/write_chat(target, text)
+	if(disabled)
+		return
+	var/target_text = ""
+	if(target == GLOB.clients)
+		target_text = "world"
+	else if(islist(target))
+		var/list/target_keys = list()
+		for(var/T in target)
+			var/client/C = CLIENT_FROM_VAR(T)
+			if(C)
+				target_keys += C.ckey
+		if(!length(target_keys))
+			return
+		target_text = jointext(target_keys, ",")
+	else
+		var/client/C = CLIENT_FROM_VAR(target)
+		if(C)
+			target_text = C.ckey
+		else
+			return
+	var/json_encoded = json_encode(text)
+	write_event_line("chat [target_text] [last_chat_message == json_encoded ? "=" : json_encoded]")
+	last_chat_message = json_encoded
+
+// use for stuff that should appear in the main/targetless chat panel
+/datum/controller/subsystem/demo/proc/write_chat_global(text)
+	if(disabled)
+		return
+	var/json_encoded = json_encode(text)
+	write_event_line("chat global [last_chat_message == json_encoded ? "=" : json_encoded]")
+	last_chat_message = json_encoded
```

Comment:
this needs https://github.com/yogstation13/demo-viewer/pull/42 in order to properly show in the viewer, but it won't viewers without it, it just won't show the global messages.